docs/guides/transfer_learning.ipynb

```diff
@@ -0,0 +1,308 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Transfer Learning"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this guide we will walk through the process of doing Transfer Learning in Flax. There are various ways to do Transfer Learning and the best approach is task dependent, these range from just using the pretrained model as a feature extractor to fine-tuning the entire model.\n",
+    "\n",
+    "We will go through the main steps needed to perform Transfer Learning for a simple classification scenario which include:\n",
+    "* Loading a pretrained model\n",
+    "* Extracting a submodule from the pretrained model\n",
+    "* Performing 'parameter surgery' to transfer the pretrained parameters into the new model structure.\n",
+    "* Defining an optimization procedure.\n",
+    "\n",
+    "#### Setup"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": [
+     "skip-execution"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "! pip install transformers[flax]\n",
+    "! pip install -U jax jaxlib"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Creating the model\n",
+    "We will start by creating a `load_model` function that will return a Flax Module and its pretrained variables. In this specific example we'll be using HuggingFace's `FlaxCLIPVisionModel` model, which will extract a `FlaxCLIPModule` module."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%capture\n",
+    "from IPython.display import clear_output\n",
+    "from transformers import FlaxCLIPModel\n",
+    "\n",
+    "# Note: FlaxCLIPModel is not a Flax Module\n",
+    "def load_model():\n",
+    "  clip = FlaxCLIPModel.from_pretrained('openai/clip-vit-base-patch32')\n",
+    "  clear_output(wait=False) # clear the loading messages\n",
+    "  module = clip.module # extract the Flax Module\n",
+    "  variables = {'params': clip.params}\n",
+    "  return module, variables"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Extracting a submodule\n",
+    "Calling `load_model` gives us the `FlaxCLIPModule` which is composed of a text and vision submodules. Let suppose we want the `vision_model` submodule defined inside `.setup()`, to do this we we can use `bind` to force `.setup()` to run (see [The Module lifecycle](https://flax.readthedocs.io/en/latest/advanced_topics/module_lifecycle.html#setup)) and grant us access to submodules and their variables."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "clip_module, clip_variables = load_model()\n",
+    "# bind variables, this runs .setup() making .vision_model available\n",
+    "bounded_clip = clip_module.bind(clip_variables)\n",
```

Comment:
bounded_clip is leaked here together with the variables. You could also write:
`model, model_vars = nn.apply(lambda clip: (clip.vision_model.clone(), clip.vision_model.variables), clip_module)(clip_variables)`

Reply:
I like it! Originally I had some `del`s but it looked awkward some I took them out.